warp10/src/main/java/io/warp10/standalone/StandaloneDirectoryClient.java

```diff
@@ -680,19 +706,23 @@ public void register(Metadata metadata) throws IOException {
     }
   }
 
-  public synchronized void unregister(Metadata metadata) {
-    if (!classids.containsKey(metadata.getName())) {
-      return;
-    }
+  public void unregister(Metadata metadata) {
     // 128BITS
-    long labelsId = GTSHelper.labelsId(this.labelsLongs, metadata.getLabels());
-    if (!metadatas.get(metadata.getName()).containsKey(labelsId)) {
-      return;
-    }
-    metadatas.get(metadata.getName()).remove(labelsId);
-    if (metadatas.get(metadata.getName()).isEmpty()) {
-      metadatas.remove(metadata.getName());
-      classids.remove(metadata.getName());
+    long labelsId;
+
+    synchronized(metadatas) {
+      if (!classids.containsKey(metadata.getName())) {
+        return;
+      }
+      labelsId = GTSHelper.labelsId(this.labelsLongs, metadata.getLabels());
```

Comment:
Actually second thoughts about where labelsId is computed. If it is occasional to call unregister for a GTS whose class is  no longer known it might be better to always compute labelsId where labelsId is declared rather than in the synchronized block since the computation is CPU heavy and therefore would extend the time spent in the syncrhonized block which would block all threads attempting to synchronize on metadatas. So probably better to always compute labelsId and do so outside of the synchronized block.